Modify linked list

cmchan04 opened the pull request:
Changed the ```Transaction``` object in linked list to ```Transaction*```pointer for better performance.
Update the function in ```main``` to avoid duplicate local copy.
Rearrange some of the ```copyArray```or linked list to increase fairness for comparison. 

I got a more reasonable results (like beginning for current process is similar to previous final) and slightly increase the performance for linked list (on my side at least).

Do take a look when you are free, @BengRhui . I not neijuan ah, I slept early and just wake up.🫡🫡 

For the documentation, I will be able to complete it on Sunday but at most on Monday. Cuz need some modification on what I have written.

cmchan04 committed:
Convert Transaction Object to pointer in linked list

cmchan04 committed:
Resolve main

cmchan04 committed:
Update gitignore to avoid accidentally pushed json

BengRhui commented:
Yes meh, not neijuan meh 🫤 But anyways let me perform some testing and see if the changes have led to any invalid results. I'll update here if there's anything ya

BengRhui committed:
Add location to list print contents

BengRhui committed:
Add check for sampling

Just to make sure that when the defined sampling size is larger than the list size, no sampling takes place

BengRhui commented:
The Transaction items I used for testing:

```
    DoublyLinkedList* overallTransactionList = new DoublyLinkedList();
    string data1[] =  {"T100000", "2023-08-22T09:22:43", "ACC001", "ACC002", "343.78", "withdrawal", "utilities", "Tokyo", "mobile", "false", "", "5.2", "-0.21", "3", "0.22", "card", "13.101.214.112", "D100001"};
    string data2[] =  {"T100001", "2023-08-04T01:58:02", "ACC003", "ACC004", "419.65", "withdrawal", "online", "Toronto", "atm", "false", "", "10.1", "-0.14", "7", "0.96", "ACH", "172.52.47.194", "D100002"};
    string data3[] =  {"T100002", "2023-05-12T11:39:33", "ACC005", "ACC006", "2773.86", "deposit", "other", "London", "pos", "false", "", "0.0", "-1.78", "20", "0.89", "card", "185.98.35.23", "D100003"};
    string data4[] =  {"T100003", "2023-07-01T14:15:00", "ACC007", "ACC008", "850.00", "payment", "groceries", "New York", "mobile", "false", "", "2.3", "0.11", "4", "0.45", "upi", "104.26.10.78", "D100004"};
    string data5[] =  {"T100004", "2023-07-02T10:00:00", "ACC009", "ACC010", "1299.99", "withdrawal", "electronics", "Singapore", "web", "true", "phishing", "0.9", "0.02", "5", "1.23", "wire_transfer", "203.0.113.1", "D100005"};
    string data6[] =  {"T100005", "2023-06-15T08:45:30", "ACC011", "ACC012", "33.20", "transfer", "healthcare", "Bangkok", "mobile", "false", "", "1.0", "0.12", "2", "0.76", "card", "198.51.100.22", "D100006"};
    string data7[] =  {"T100006", "2023-08-10T13:00:00", "ACC013", "ACC014", "1700.00", "deposit", "shopping", "Paris", "pos", "true", "stolen_card", "0.0", "1.50", "9", "1.01", "ACH", "8.8.8.8", "D100007"};
    string data8[] =  {"T100007", "2023-06-20T18:30:00", "ACC015", "ACC016", "95.45", "payment", "groceries", "Berlin", "atm", "false", "", "3.2", "-0.50", "3", "0.88", "upi", "1.1.1.1", "D100008"};
    string data9[] =  {"T100008", "2023-07-22T12:00:00", "ACC017", "ACC018", "289.00", "withdrawal", "online", "Seoul", "web", "true", "malware", "7.5", "0.45", "6", "1.45", "wire_transfer", "192.0.2.33", "D100009"};
    string data10[] = {"T100009", "2023-08-11T17:15:00", "ACC019", "ACC020", "510.75", "transfer", "utilities", "Dubai", "mobile", "false", "", "1.9", "-0.10", "5", "0.65", "card", "203.0.113.5", "D100010"};
    string data11[] = {"T100010", "2023-08-12T08:00:00", "ACC021", "ACC022", "760.80", "deposit", "education", "Osaka", "atm", "false", "", "4.0", "0.55", "3", "0.91", "wire_transfer", "192.168.1.1", "D100011"};
    string data12[] = {"T100011", "2023-07-30T23:59:59", "ACC023", "ACC024", "620.00", "withdrawal", "fashion", "Manila", "pos", "true", "fake_identity", "5.3", "-1.1", "7", "1.35", "upi", "10.0.0.1", "D100012"};
    string data13[] = {"T100012", "2023-06-01T16:45:00", "ACC025", "ACC026", "45.90", "payment", "food", "Delhi", "mobile", "false", "", "0.0", "-0.09", "2", "0.32", "wire_transfer", "203.0.113.10", "D100013"};
    string data14[] = {"T100013", "2023-05-20T10:10:10", "ACC027", "ACC028", "988.88", "transfer", "others", "Rome", "web", "true", "unauthorized_access", "8.8", "0.0", "10", "1.78", "wire_transfer", "192.0.2.20", "D100014"};
    string data15[] = {"T100014", "2023-04-01T01:23:45", "ACC029", "ACC030", "1550.00", "deposit", "services", "Kuala Lumpur", "atm", "false", "", "2.2", "0.25", "4", "0.50", "ACH", "172.16.254.1", "D100015"};
    string data16[] = {"T100015", "2023-03-15T15:15:15", "ACC031", "ACC032", "48.99", "withdrawal", "groceries", "Hanoi", "mobile", "false", "", "0.5", "-0.30", "2", "0.73", "upi", "198.51.100.50", "D100016"};
    string data17[] = {"T100016", "2023-02-28T14:00:00", "ACC033", "ACC034", "134.00", "payment", "travel", "Jakarta", "pos", "false", "", "3.1", "0.65", "6", "0.97", "card", "8.8.4.4", "D100017"};
    string data18[] = {"T100017", "2023-01-01T00:00:01", "ACC035", "ACC036", "999.99", "transfer", "electronics", "Sydney", "web", "true", "data_breach", "6.6", "1.15", "8", "1.61", "wire_transfer", "203.0.113.15", "D100018"};
    string data19[] = {"T100018", "2023-07-10T09:30:00", "ACC037", "ACC038", "222.22", "deposit", "entertainment", "Bangkok", "mobile", "false", "", "0.0", "-0.17", "5", "0.40", "ACH", "192.168.100.1", "D100019"};
    string data20[] = {"T100019", "2023-06-25T11:00:00", "ACC039", "ACC040", "875.50", "withdrawal", "shopping", "Hong Kong", "atm", "false", "", "2.7", "0.34", "3", "0.83", "upi", "203.0.113.25", "D100020"};

    Transaction t1(data1), t2(data2), t3(data3), t4(data4), t5(data5),
                t6(data6), t7(data7), t8(data8), t9(data9), t10(data10),
                t11(data11), t12(data12), t13(data13), t14(data14), t15(data15),
                t16(data16), t17(data17), t18(data18), t19(data19), t20(data20);

    overallTransactionList -> insertAtEnd(&t1);
    overallTransactionList->insertAtEnd(&t2);
    overallTransactionList->insertAtEnd(&t3);
    overallTransactionList->insertAtEnd(&t4);
    overallTransactionList->insertAtEnd(&t5);
    overallTransactionList->insertAtEnd(&t6);
    overallTransactionList->insertAtEnd(&t7);
    overallTransactionList->insertAtEnd(&t8);
    overallTransactionList->insertAtEnd(&t9);
    overallTransactionList->insertAtEnd(&t10);
    overallTransactionList->insertAtEnd(&t11);
    overallTransactionList->insertAtEnd(&t12);
    overallTransactionList->insertAtEnd(&t13);
    overallTransactionList->insertAtEnd(&t14);
    overallTransactionList->insertAtEnd(&t15);
    overallTransactionList->insertAtEnd(&t16);
    overallTransactionList->insertAtEnd(&t17);
    overallTransactionList->insertAtEnd(&t18);
    overallTransactionList->insertAtEnd(&t19);
    overallTransactionList->insertAtEnd(&t20);
```

The testing looks fine, but let me try double check if there's any changes to make.

BengRhui committed:
Remove const for transactions

Instead of using const_cast<>, i think it'll be better if we just remove the const on the transaction objects. Much safer if compared to const_cast.

BengRhui committed:
Revert "Remove const for transactions"

This reverts commit 76af115f62af41b725254b518d6d5a8271cf7070.

BengRhui committed:
Declare transaction object as const

This avoids modifying the remaining code with const keyword

BengRhui committed:
Revert "Declare transaction object as const"

This reverts commit dfc0eb7519096bb1fd43267659ddc8db1b19037e.

BengRhui committed:
Reapply "Remove const for transactions"

This reverts commit ed673a05577b5997b19ea0728d83d3c001cc976a.

BengRhui commented:
@cmchan04 I think we shall try avoid using `const_cast<>`, coz:

```
Great question — and no, you should not use const_cast<Transaction*>(data) here unless you absolutely need to modify the object that was originally declared as const. It's generally unsafe unless you're very sure the object isn't actually const.

🔎 Let’s break it down:

Node::Node(const Transaction* data, const int index) {
    if (data == nullptr) throw invalid_argument("Data cannot be null.");
    this->transactionObject = const_cast<Transaction*>(data);
}

Here:

You're receiving a const Transaction* (i.e., a pointer to an object that shouldn't be modified).
Then you're removing the const-ness with const_cast<Transaction*>(data) so you can store a mutable pointer.

❗ Why this is dangerous:
If data actually points to a const Transaction object (i.e., one that was created as const), and later you modify transactionObject, it causes undefined behavior.

const Transaction tx = ...;
Node node(&tx, 0);  // UB: you're casting away const from a genuinely const object

✅ Better and safer alternatives:

Option 1: Accept Transaction* if you intend to modify it

If you need to store a non-const pointer, change the parameter to:

Node::Node(Transaction* data, const int index) {
    if (data == nullptr) throw invalid_argument("Data cannot be null.");
    this->transactionObject = data;
}
This way, only non-const objects can be passed, and no cast is needed.

Option 2: Store const Transaction* if the node is read-only

If your Node class does not modify the transaction, it’s best to keep the pointer as const:

const Transaction* transactionObject;
Then your constructor becomes clean and cast-free:

Node::Node(const Transaction* data, const int index) {
    if (data == nullptr) throw invalid_argument("Data cannot be null.");
    this->transactionObject = data;
}

✅ Final Recommendation:
Unless you truly must modify the transaction through the node (which is rare in immutable designs), prefer:

const Transaction* transactionObject;

And avoid const_cast completely — it’s often a red flag for unsafe code unless used with extreme care.

Let me know if your list intends to modify transactionObject, and I can help you design accordingly.
```

BengRhui commented:
I tried to just declare the `Transaction` object as `const`, but it leads to errors (hence the revert of commits haiya), to make things safer I guess we shall just remove the `const` keyword from all the `Transaction` object, wdyt @cmchan04?

Psst: Btw let me try run the program one more time and see if everything works well, should have no issues alrdy (I guess?)

cmchan04 commented:
> I tried to just declare the `Transaction` object as `const`, but it leads to errors (hence the revert of commits haiya), to make things safer I guess we shall just remove the `const` keyword from all the `Transaction` object, wdyt @cmchan04?
> 
> Psst: Btw let me try run the program one more time and see if everything works well, should have no issues alrdy (I guess?)

OK you test first, I add some other measurements in tracker, like (second, MB, GB)

cmchan04 committed:
Add MB to performance tracker

cmchan04 committed:
Merge branch 'Modify_Linked_List' of https://github.com/cmchan04/DSTR_Assignment into Modify_Linked_List

BengRhui reviewed (approved):
Should be all good as long as the no significant changes are made towards the logic of the code

cmchan04 commented:
I will merge it to main, if anything happens we still have the backup branch
